Wrong pid on windows: Error: kill ESRCH

Every new started process gets wrong pid which is different from ones shown in `task manager`.
Could be reproduced with the [demo of xterm.js](https://github.com/sourcelair/xterm.js/tree/bebe43e5e108b081a8ead57b881e7ea1504164d7/demo).
Related to [xterm.js issue](https://github.com/sourcelair/xterm.js/pull/584).

```sh
> xterm@2.3.2 start C:\Users\coderaiser\Documents\GitHub\xterm.js
> node demo/app

App listening to http://127.0.0.1:3000
Created terminal with PID: 928
Connected to terminal 928
internal/process.js:173
      throw errnoException(err, 'kill');
      ^

Error: kill ESRCH
    at exports._errnoException (util.js:1022:11)
    at process.kill (internal/process.js:173:13)
    at WebSocket.<anonymous> (C:\Users\coderaiser\Documents\GitHub\xterm.js\demo\app.js:72:13)
    at emitTwo (events.js:111:20)
    at WebSocket.emit (events.js:191:7)
    at WebSocket.cleanupWebsocketResources (C:\Users\coderaiser\Documents\GitHub\xterm.js\node_modules\ws\lib\WebSocket.js:950:8)
    at emitOne (events.js:101:20)
    at Socket.emit (events.js:188:7)
    at emitErrorNT (net.js:1281:8)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)

npm ERR! Windows_NT 10.0.14393
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "start"
npm ERR! node v7.2.0
npm ERR! npm  v3.10.9
npm ERR! code ELIFECYCLE
npm ERR! xterm@2.3.2 start: `node demo/app`
npm ERR! Exit status 1
```

--- a/src/win/pty.cc
+++ b/src/win/pty.cc
@@ -79,6 +79,7 @@ static bool remove_pipe_handle(int handle) {
     winpty_t *ptyHandle = ptyHandles[i];
     if((int)winpty_agent_process(ptyHandle) == handle) {
       winpty_free(ptyHandle);
+      ptyHandles.erase(ptyHandles.begin() + i);
       ptyHandle = nullptr;
       return true;
     }

--- a/src/win/pty.cc
+++ b/src/win/pty.cc
@@ -317,11 +317,11 @@ static NAN_METHOD(PtyKill) {
     || !info[0]->IsNumber() // pid
     || !info[1]->IsNumber()) // innerPidHandle
   {
-    return Nan::ThrowError("Usage: pty.kill(pid)");
+    return Nan::ThrowError("Usage: pty.kill(pid, innerPidHandle)");
   }
 
   int handle = info[0]->Int32Value();
-  HANDLE innerPidHandle = (HANDLE)info[0]->Int32Value();
+  HANDLE innerPidHandle = (HANDLE)info[1]->Int32Value();
 
   winpty_t *pc = get_pipe_handle(handle);
 

--- a/src/win/pty.cc
+++ b/src/win/pty.cc
@@ -246,6 +246,8 @@ open:
   // Pty object values.
   Local<Object> marshal = Nan::New<Object>();
 
+  marshal->Set(Nan::New<String>("innerPid").ToLocalChecked(), Nan::New<Number>((int)GetProcessId(handle)));
+  marshal->Set(Nan::New<String>("innerPidHandle").ToLocalChecked(), Nan::New<Number>((int)handle));
   marshal->Set(Nan::New<String>("pid").ToLocalChecked(), Nan::New<Number>((int)winpty_agent_process(pc)));
   marshal->Set(Nan::New<String>("pty").ToLocalChecked(), Nan::New<Number>(InterlockedIncrement(&ptyCounter)));
   marshal->Set(Nan::New<String>("fd").ToLocalChecked(), Nan::New<Number>(-1));
@@ -311,18 +313,21 @@ static NAN_METHOD(PtyResize) {
 static NAN_METHOD(PtyKill) {
   Nan::HandleScope scope;
 
-  if (info.Length() != 1
-    || !info[0]->IsNumber()) // pid
+  if (info.Length() != 2
+    || !info[0]->IsNumber() // pid
+    || !info[1]->IsNumber()) // innerPidHandle
   {
     return Nan::ThrowError("Usage: pty.kill(pid)");
   }
 
   int handle = info[0]->Int32Value();
+  HANDLE innerPidHandle = (HANDLE)info[0]->Int32Value();
 
   winpty_t *pc = get_pipe_handle(handle);
 
   assert(pc != nullptr);
   assert(remove_pipe_handle(handle));
+  CloseHandle(innerPidHandle);
 
   return info.GetReturnValue().SetUndefined();
 }

--- a/src/win/pty.cc
+++ b/src/win/pty.cc
@@ -36,20 +36,6 @@ extern "C" void init(Handle<Object>);
 static std::vector<winpty_t *> ptyHandles;
 static volatile LONG ptyCounter;
 
-struct winpty_s {
-  winpty_s();
-  HANDLE controlPipe;
-  HANDLE coninPipeName;
-  HANDLE conoutPipeName;
-};
-
-winpty_s::winpty_s() :
-  controlPipe(nullptr),
-  coninPipeName(nullptr),
-  conoutPipeName(nullptr)
-{
-}
-
 /**
 * Helpers
 */

--- a/src/windowsPtyAgent.ts
+++ b/src/windowsPtyAgent.ts
@@ -21,6 +21,9 @@ export class WindowsPtyAgent {
   private _inSocket: net.Socket;
   private _outSocket: net.Socket;
   private _pid: number;
+  private _innerPid: number;
+  private _innerPidHandle: number;
+  
   private _fd: any;
   private _pty: number;
 
@@ -50,6 +53,8 @@ export class WindowsPtyAgent {
 
     // Terminal pid.
     this._pid = term.pid;
+    this._innerPid = term.innerPid;
+    this._innerPidHandle = term.innerPidHandle;
 
     // Not available on windows.
     this._fd = term.fd;
@@ -83,7 +88,7 @@ export class WindowsPtyAgent {
     this._inSocket.writable = false;
     this._outSocket.readable = false;
     this._outSocket.writable = false;
-    pty.kill(this.pid);
+    pty.kill(this.pid, this._innerPidHandle);
   }
 }
 

--- a/src/windowsPtyAgent.ts
+++ b/src/windowsPtyAgent.ts
@@ -23,7 +23,7 @@ export class WindowsPtyAgent {
   private _pid: number;
   private _innerPid: number;
   private _innerPidHandle: number;
-  
+
   private _fd: any;
   private _pty: number;
 

